### ISSUE TITLE
Compilation Error in Visual Studio 2019

### ISSUE THREAD

**Issue by gamagan** [janet-lang/janet#358]
I checked out the latest version of master, as of 4/25/20.
In my code (c++), I included "janet.h" and attempted to compile the code, but the compilation failed:

```
1>Z:\ThirdParty\Include\janet/janet.h(462,37): error C2526: 'janet_wrap_nil': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(463,44): error C2526: 'janet_wrap_number': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(464,38): error C2526: 'janet_wrap_true': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(465,39): error C2526: 'janet_wrap_false': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(466,42): error C2526: 'janet_wrap_boolean': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(467,52): error C2526: 'janet_wrap_string': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(468,52): error C2526: 'janet_wrap_symbol': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(469,53): error C2526: 'janet_wrap_keyword': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(470,48): error C2526: 'janet_wrap_array': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(471,49): error C2526: 'janet_wrap_tuple': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(472,52): error C2526: 'janet_wrap_struct': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(473,48): error C2526: 'janet_wrap_fiber': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(474,50): error C2526: 'janet_wrap_buffer': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(475,54): error C2526: 'janet_wrap_function': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(476,55): error C2526: 'janet_wrap_cfunction': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(477,48): error C2526: 'janet_wrap_table': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(478,45): error C2526: 'janet_wrap_abstract': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(479,44): error C2526: 'janet_wrap_pointer': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(480,46): error C2526: 'janet_wrap_integer': C linkage function cannot return C++ class 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(313): message : see declaration of 'Janet'
1>Z:\ThirdParty\Include\janet/janet.h(779,23): warning C4200: nonstandard extension used: zero-sized array in struct/union
1>Z:\ThirdParty\Include\janet/janet.h(779,23): message : This member will be ignored by a defaulted constructor or copy/move assignment operator
1>Z:\ThirdParty\Include\janet/janet.h(788,25): warning C4200: nonstandard extension used: zero-sized array in struct/union
1>Z:\ThirdParty\Include\janet/janet.h(788,25): message : This member will be ignored by a defaulted constructor or copy/move assignment operator
1>Z:\ThirdParty\Include\janet/janet.h(796,25): warning C4200: nonstandard extension used: zero-sized array in struct/union
1>Z:\ThirdParty\Include\janet/janet.h(796,25): message : This member will be ignored by a defaulted constructor or copy/move assignment operator
1>Z:\ThirdParty\Include\janet/janet.h(804,21): warning C4200: nonstandard extension used: zero-sized array in struct/union
1>Z:\ThirdParty\Include\janet/janet.h(804,21): message : This member will be ignored by a defaulted constructor or copy/move assignment operator
1>Z:\ThirdParty\Include\janet/janet.h(868,25): warning C4200: nonstandard extension used: zero-sized array in struct/union
1>Z:\ThirdParty\Include\janet/janet.h(868,25): message : This member will be ignored by a defaulted constructor or copy/move assignment operator

```
The lines around janet.h:313 are like this:

```
#ifdef JANET_NANBOX_64
typedef union Janet Janet;  // Line 313
#elif defined(JANET_NANBOX_32)
```

**Comment by bakpakin** [comment]
Not sure what you are trying to do here, but some compiler flags/indication of what compiler is being used would be nice. It is almost an issue with your configuration, as we use visual studio 2019 and MSVC in the windows CI to build janet for windows. Did you add janetconf.h to your project as well?

**Comment by gamagan** [comment]
ok

**Comment by bakpakin** [comment]
So this turned out to be much easier to fix than anticipated, fixed in 57332c5ccfe17effad8aace28931dd76dbeeaf97 by changing the order of declarations.

It seems it is a (from my point of view) rather arbitrary limit of MSVC - gcc/g++ works fine for this, and clang/clang++ as well, although they do emit a warning.
